### PR TITLE
2 - Basic Weapon Controller

### PR DESCRIPTION
**Pull request by overseer25**
Basic implementation of the Weapon Controller completed. Below are the following changes made:

- A Weapon base class, and two mock weapons for testing.
- A Weapon Slot class that represents a slot in the player's weapon wheel. Has a weapon as a child object.
- A Weapon Controller class. Has features such as "Adding" a weapon to the wheel and swapping currently selected weapons.
- Utility script folder, and added a Tag utility script that we can use to reference tags we create from one place, rather than hardcoding strings everywhere.
- A Collectable interface and CollectableWeapon class that implements it. This allows for physical objects in the game world to be used to add weapons/ammo to the player's inventory.
- Added my dev scene folder with an example scene of the basic weapon controller working.

I need to commit more often...

### CODE REVIEW

File: Assets/Scripts/Collectables/CollectableWeapon.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+    private void OnTriggerEnter(Collider other)
+    {
+        switch(other.tag)
+        {
+            case Tags.PLAYER_TAG:
+                WeaponController weapControl = other.GetComponent<WeaponController>();
+                if (weapControl != null)
+                    weapControl.AddWeapon(this);
+                Destroy(gameObject);
```
Comment (packerbacker8) [inline Assets/Scripts/Collectables/CollectableWeapon.cs]: We discussed this already but we may want to not destroy these objects. Object pooling across scenes might be something we consider in the future.

File: Assets/Scripts/Player/WeaponController.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            return;
+
+        weaponWheel[currentSlot].Deactivate();
+        weaponWheel[index].Activate();
+        currentSlot = index;
+    }
+
+    /// <summary>
+    /// Add a weapon to the player. If the weapon already exists in the player's inventory, and cannot be dual-wielded (or already is), gives ammo instead, defined by the weapon's script.
+    /// </summary>
```
Comment (packerbacker8) [inline Assets/Scripts/Player/WeaponController.cs]: Probably can make this a multi line comment. Styling or whatever XD

File: Assets/Scripts/Player/WeaponController.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    /// <param name="weapon"></param>
+    public void AddWeapon(CollectableWeapon weapon)
+    {
+        if (weapon == null)
+            return;
```
Comment (packerbacker8) [inline Assets/Scripts/Player/WeaponController.cs]: Do we want any logging here asking why it is null?

File: Assets/Scripts/Player/WeaponController.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                currentSlot = i;
+                break;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Switch from the currently selected weapon slot to the provided weapon slot, using the index.
+    /// </summary>
+    /// <param name="index"></param>
```
Comment (packerbacker8) [inline Assets/Scripts/Player/WeaponController.cs]: Loving the doc comments (I am not doing enough of them myself, gotta add them. But do we want to describe the params?

File: Assets/Scripts/Player/WeaponSlot.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+{
+    /// <summary>
+    /// Can the weapon in this slot be dual wielded?
+    /// </summary>
+    public bool canDualWield;
+
+    /// <summary>
+    /// Does the slot have two of this weapon (only set if the weapon can be dual-wielded)?
+    /// </summary>
+    public bool isDualWielded;
```
Comment (packerbacker8) [inline Assets/Scripts/Player/WeaponSlot.cs]: Will a dual wielded weapon be a graphical only thing or will we be using two gameobjects to represent it?

File: Assets/Scripts/Weapons/WeaponBase.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    /// </summary>
+    public void Disable()
+    {
+        gameObject.SetActive(false);
+    }
+
+    /// <summary>
+    /// Calling this method causes the weapon to fire.
+    /// </summary>
+    public abstract void Fire();
```
Comment (packerbacker8) [inline Assets/Scripts/Weapons/WeaponBase.cs]: With the possibility for melee weapons (like that whip/hook thing we discussed) and maybe others, do we want this to be called attack or something to include those or just understand naming might be a bit odd when it comes to melee? or have them have a different base entirely and just make sure they fit into weapon slots like the current system? or have a whole different area they reside and they kind of are unlocked in a separate fashion to guns?

File: Assets/Scripts/Player/WeaponController.cs (hunk 3 above)
Comment (overseer25) [reply Assets/Scripts/Player/WeaponController.cs]: Yes, but I don't have access to the logger you are working on. I can add a simple Debug.Log for now.

File: Assets/Scripts/Player/WeaponController.cs (hunk 4 above)
Comment (overseer25) [reply Assets/Scripts/Player/WeaponController.cs]: Yeah, should probably do that. I'll add that to all the xml comments I've made.

File: Assets/Scripts/Player/WeaponSlot.cs (hunk 5 above)
Comment (overseer25) [reply Assets/Scripts/Player/WeaponSlot.cs]: I haven't quite decided that yet. I was just adding the bools for future use (probably not a good idea to add code for stuff that I am not working on yet).

I imagine dual-wielded weapons will be a separate game object from the single-wielded variant, and have it's own models.

File: Assets/Scripts/Weapons/WeaponBase.cs (hunk 6 above)
Comment (overseer25) [reply Assets/Scripts/Weapons/WeaponBase.cs]: If it is confusing, I can create two additional base classes, MeleeWeaponBase and ProjectileWeaponBase, that both inherit from the WeaponBase class. These base classes can then define separate functionality.

File: Assets/Scripts/Collectables/CollectableWeapon.cs (hunk 1 above)
Comment (overseer25) [reply Assets/Scripts/Collectables/CollectableWeapon.cs]: Yeah, I think a "Collectable" object pool is something that can be added later.

File: Assets/Scripts/Player/WeaponController.cs (hunk 3 above)
Comment (packerbacker8) [reply Assets/Scripts/Player/WeaponController.cs]: This should be captured by my console logging thing when it gets added. As in using Debug.Log will show up in the in game console without needing to switch anything out.

File: Assets/Scripts/Player/WeaponSlot.cs (hunk 5 above)
Comment (packerbacker8) [reply Assets/Scripts/Player/WeaponSlot.cs]: Seems reasonable to me.

File: Assets/Scripts/Weapons/WeaponBase.cs (hunk 6 above)
Comment (packerbacker8) [reply Assets/Scripts/Weapons/WeaponBase.cs]: I am not sure it was confusing, just something to consider as we move forward. Using the method name Fire for melee weapons would work fine. Just something i was considering.

File: Assets/Scripts/Player/WeaponSlot.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+            return ammoCount;
+        }
+        set
+        {
+            if (weapon == null)
+                return;
+            ammoCount = value;
+
+            // Prevent ammo overflow
+            if (ammoCount > maxAmmo)
```
Comment (kucab2345) [inline Assets/Scripts/Player/WeaponSlot.cs]: Would recommend using a clamp on the ammoCount...

ammoCount = Mathf.Clamp(ammoCount, 0, maxAmmo)

Less complexity, intention would be clearer.

**Pull request merged**